File: flutter/feature_core/lib/src/core_feature_module.dart
import 'dart:async';
import 'package:contracts/contracts.dart';
import 'ui/login/login_screen.dart';
import 'ui/splash/splash_screen.dart';

/// Core infrastructure module — handles authentication UI and app entry points.
///
/// Provides [Routes.splash] and [Routes.login] (always available, no role
/// restriction). Listens for [UserLoggedOutEvent] to redirect back to login.
///
/// Priority 1000 — initialized first so navigation entry point is ready
/// before any feature module routes are registered.
///
/// Mirrors Android [CoreFeatureModule].
class CoreFeatureModule extends AppModule {
  @override
  final ModuleMetadata metadata = const ModuleMetadata(
    id: 'core',
    name: 'Core',
    version: '1.0.0',
    requiredRoles: {Role.ADMIN, Role.STAFF, Role.CUSTOMER, Role.GUEST},
    priority: 1000,
  );

  StreamSubscription<UserLoggedOutEvent>? _logoutSub;

  @override
  void initialize(ModuleContext context) {
    _logoutSub = context.eventBus
        .on<UserLoggedOutEvent>()
        .listen((_) => context.navigate(Routes.login));
  }

  @override
  List<ModuleRoute> provideRoutes() => [
        ModuleRoute(
          route: Routes.splash,
          requiredRole: Role.GUEST,
          builder: (context) => const SplashScreen(),
        ),
        ModuleRoute(
          route: Routes.login,
          requiredRole: Role.GUEST,
          builder: (context) => const LoginScreen(),
        ),
      ];

  @override
  List<UISlot> provideWidgets() => [];

  @override
  void onDestroy() {
    _logoutSub?.cancel();
    _logoutSub = null;
  }
}

File: android/feature-core/src/main/java/com/densitech/largescale/feature/core/CoreFeatureModule.kt
package com.densitech.largescale.feature.core

import com.densitech.largescale.contracts.AppModule
import com.densitech.largescale.contracts.ModuleContext
import com.densitech.largescale.contracts.ModuleMetadata
import com.densitech.largescale.contracts.ModuleRoute
import com.densitech.largescale.contracts.Role
import com.densitech.largescale.contracts.Routes
import com.densitech.largescale.contracts.UISlot
import com.densitech.largescale.contracts.UserLoggedOutEvent
import com.densitech.largescale.contracts.on

/**
 * Core infrastructure module — handles authentication UI and app entry points.
 *
 * Provides: SPLASH and LOGIN routes (always available, no role restriction).
 * Listens for [UserLoggedOutEvent] to redirect back to login.
 *
 * Priority 1000 — initialized first so navigation entry point is always ready.
 */
class CoreFeatureModule : AppModule {

    override val metadata = ModuleMetadata(
        id = "core",
        name = "Core",
        version = "1.0.0",
        requiredRoles = setOf(Role.ADMIN, Role.STAFF, Role.CUSTOMER, Role.GUEST),
        priority = 1000
    )

    private var moduleContext: ModuleContext? = null

    override fun initialize(context: ModuleContext) {
        moduleContext = context

        // On logout, navigate back to login and clear backstack
        context.eventBus.on<UserLoggedOutEvent> {
            context.navigate(Routes.LOGIN)
        }
    }

    override fun provideRoutes() = listOf(
        ModuleRoute(route = Routes.SPLASH, requiredRole = Role.GUEST),
        ModuleRoute(route = Routes.LOGIN,  requiredRole = Role.GUEST)
    )

    override fun provideWidgets(): List<UISlot> = emptyList()

    override fun onDestroy() {
        moduleContext = null
    }
}
